collection-scripts/gather-sosreports

```diff
@@ -0,0 +1,132 @@
+#!/bin/bash
+
+BASE_COLLECTION_PATH="/must-gather"
+SOSREPORT_DUMP_PATH=${OUT:-"${BASE_COLLECTION_PATH}/node_sosreports"}
+
+function gather_sosreport() {
+	local node=$1
+	local registry=$2
+	local image=$3
+
+	tmp_file=$(mktemp)
+	cat <<'EOF' > $tmp_file
+#!/bin/bash
+
+chroot /host /usr/bin/sudo --preserve-env /bin/bash -x <<'EOT'
+# defaults first
+REGISTRY=registry.redhat.io
+IMAGE=rhel8/support-tools
+TOOLBOX_NAME=toolbox-"$(whoami)-sosreport-$(uuidgen | cut -b-8)"
+TOOLBOXRC="/root/.toolboxrc"
+
+# second if .toolboxrc file exists, take its values
+if [ -f "${TOOLBOXRC}" ]; then
+    echo ".toolboxrc file detected, overriding defaults..."
+    source "${TOOLBOXRC}"
+fi
+
+# third if user overrides are set, use those
+USER_REGISTRY=""
+USER_IMAGE=""
+if [ "${USER_REGISTRY}" != "" ]; then
+    REGISTRY="${USER_REGISTRY}"
+fi
+if [ "${USER_IMAGE}" != "" ]; then
+    IMAGE="${USER_IMAGE}"
+fi
+
+# now, concat registry and image
+TOOLBOX_IMAGE="${REGISTRY}"/"${IMAGE}"
+
+podman pull --authfile /var/lib/kubelet/config.json "$TOOLBOX_IMAGE"
+
+podman run \
+  --rm \
+  --hostname $(hostname) \
+  --name "$TOOLBOX_NAME" \
+  --privileged \
+  --net=host \
+  --pid=host \
+  --ipc=host \
+  -e HOST=/host \
+  -e NAME="$TOOLBOX_NAME" \
+  -e IMAGE="$IMAGE" \
+  --security-opt label=disable \
+  --volume /run:/run \
+  --volume /var/log:/var/log \
+  --volume /etc/machine-id:/etc/machine-id \
+  --volume /etc/localtime:/etc/localtime \
+  --volume /:/host \
+   "$TOOLBOX_IMAGE" \
+  sosreport -k crio.all=on -k crio.logs=on --batch --all-logs 2>/dev/null | tee /tmp/must_gather_sosreport.log
+EOT
+EOF
+	sed -i "s#^USER_REGISTRY=#USER_REGISTRY=\"$registry\"#" ${tmp_file}
+	sed -i "s#^USER_IMAGE=#USER_IMAGE=\"$image\"#" ${tmp_file}
+
+	# Get debug pod's name and exit right away
+	debug_pod=$(oc debug --to-namespace=default ${node} -o jsonpath='{.metadata.name}')
```

Comment:
We don't want to spin this pod up in the 'default namespace'. Every must-gather invocation runs in its own namespaces; and we will want to launch this pod in that namespaces as well.

Reply:
Hi,

>> I think we need to review this a bit more; @andreaskaris can you outline the motivation for the need of for sosreports exposed in this manner?

This is a customer ask, coming from a high priority customer.  https://issues.redhat.com/browse/RFE-1810 contains more details and the reason behind the ask.

>> In general; you should be using must-gather to look at cluster issues; and only using sosreport to get a 'cluster back to functional' (so that you could run a must-gather)

In the field, that unfortunately does not work, at least for our enhanced support teams. We often ask our customers to generate sosreports to get node local events and configuration (imagine SR-IOV issues, CNV issues) in additiona to must-gather reports.  So we end up asking our customers to gather a must-gather, then a specific must-gather such as CNV or SR-IOV, and in addition, we need sosreports to analyze a baremetal node's status (imagine kernel bugs ,etc.)
In the enhanced support team, we work with OpenShift in baremetal events and sosreports are the only tool that captures some of the critical data for baremetal nodes. The must-gather tool already contains a script to gather vmcores; that's a highly specialized and lowlevel operation and falls into the same category as gathering  sosreports.
Given that for our team at least it is a very common task to request sosreports on top of must-gathers, the customer must go into every single node and gather and pull a sosreport. That's cumbersome and the best tool to gather a subset of nodes is the must-gather tool.

>> (on a node where a particular issue is happening)

That's what the `-l` label invocation takes care of.

>> In  general my fear / issue with this is that we are adding a collector/collection of information that we typically reserve for situations where must-gather will not run.

As stated above, we have several high profile customers where our custom policy is: generate and gather a must-gather, and generate and gather a sosreport, even if the cluster is running. That might be a shortcoming of the other must-gather scripts, but then I strongly believe that sosreports fill a different purpose, and also one that's beyond breaking issues. Again, NFV, CNV, performance issues, kernel issues, OOM situations, all of these are things that sosreports capture and allow to analyze.

>> That and the size of must-gather will expand / grow exponentially as a result of this.

The same as the core dump collector script, I would like to add this as a standalone script which is not run by default. A user has to explicitly call the script, otherwise no sosreport will be run.

>> In general while collecting a sosreport in this way is what we document (one way we expose this), its not the recommended path to get a sosreport (the recommend path is still with ssh and starting a pod).

In order to automate this and run this via `must-gather`, I need to use `oc debug` and a toolbox container.

Reply:
> We don't want to spin this pod up in the 'default namespace'. Every must-gather invocation runs in its own namespaces; and we will want to launch this pod in that namespaces as well.

I will change this. However, then we allso must change:
https://github.com/openshift/must-gather/blob/cbf587fe6e4c97196e84aea68ad7a57210035e92/collection-scripts/gather_core_dumps#L14

Because that's where I got the `default` namespace from, in the first place :-)